pkg/spoke/addon/registration_controller.go

```diff
@@ -0,0 +1,250 @@
+	hubCSRClient   csrclient.CertificateSigningRequestInterface
+	recorder       events.Recorder
+
+	// registrationConfigs maps the addon name to a map of registrationConfigs whose key is the hash of
+	// the registrationConfig
+	addOnRegistrationConfigs map[string]map[string]registrationConfig
+}
+
+// NewAddOnRegistrationController returns an instance of addOnRegistrationController
+func NewAddOnRegistrationController(
+	clusterName string,
+	agentName string,
+	kubeconfigData []byte,
+	kubeClient kubernetes.Interface,
+	hubCSRInformer certificatesinformers.CertificateSigningRequestInformer,
+	hubAddOnInformers addoninformerv1alpha1.ManagedClusterAddOnInformer,
+	hubCSRClient csrclient.CertificateSigningRequestInterface,
+	recorder events.Recorder,
+) factory.Controller {
+	c := &addOnRegistrationController{
+		clusterName:              clusterName,
+		agentName:                agentName,
+		kubeconfigData:           kubeconfigData,
+		kubeClient:               kubeClient,
+		hubAddOnLister:           hubAddOnInformers.Lister(),
+		hubCSRInformer:           hubCSRInformer,
+		hubCSRClient:             hubCSRClient,
+		recorder:                 recorder,
+		addOnRegistrationConfigs: map[string]map[string]registrationConfig{},
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(
+			func(obj runtime.Object) string {
+				accessor, _ := meta.Accessor(obj)
+				return accessor.GetName()
+			},
+			hubAddOnInformers.Informer()).
+		WithSync(c.sync).
+		ToController("AddOnRegistrationController", recorder)
+}
+
+func (c *addOnRegistrationController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	addOnName := syncCtx.QueueKey()
+	klog.V(4).Infof("Reconciling addOn %q", addOnName)
+
+	addOn, err := c.hubAddOnLister.ManagedClusterAddOns(c.clusterName).Get(addOnName)
+	if errors.IsNotFound(err) {
+		// addon is deleted
+		return c.cleanup(ctx, addOnName)
+	}
+	if err != nil {
+		return err
+	}
+
+	// addon is deleting
+	if !addOn.DeletionTimestamp.IsZero() {
+		return c.cleanup(ctx, addOnName)
+	}
+
+	cachedConfigs := c.addOnRegistrationConfigs[addOnName]
+	configs, err := getRegistrationConfigs(addOnName, addOn.Annotations)
+	if err != nil {
+		return err
+	}
+
+	// stop registraton for the stale registration configs
+	errs := []error{}
+	for _, cachedConfig := range cachedConfigs {
+		if _, ok := configs[cachedConfig.hash]; ok {
+			continue
+		}
+
+		if err := c.stopRegistration(ctx, cachedConfig); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if err := operatorhelpers.NewMultiLineAggregate(errs); err != nil {
+		return err
+	}
+
+	for _, config := range configs {
+		cachedConfig, ok := cachedConfigs[config.hash]
+		switch {
+		case ok:
+			// get the stopFunc from the cached registration configs
+			config.stopFunc = cachedConfig.stopFunc
+		default:
+			// start registration for the new added configs
+			config.stopFunc = c.startRegistration(ctx, config)
```

Comment:
why not just put configs[config.hash] = config in the if block?

```go
if !ok {
        // start registration for the new added configs
	config.stopFunc = c.startRegistration(ctx, config)
        configs[config.hash] = config
}
```

Reply:
Rewrote in a similar way to remove the switch clause.